### PR TITLE
Feature/add configuration

### PR DESCRIPTION
**Pull request by patrick-tolosa**
Configuration Objects for Cloudinary Base.

Configurations were separated based on concerns and likelihood of change. (Cloud Configuration are global per website, while URL might not be)

### CODE REVIEW

File: __TESTS__/unit/config/cloudinaryConfig.test.ts [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+        cdn_subdomain: true,
+        cloud_name: 'cloud_name',
+        cname: 'cname',
+        api_key: 'api_key',
+        api_secret: 'api_secret',
+        secure_cdn_subdomain: true,
+        private_cdn: true,
+        secure: true,
+        secure_distribution: true,
+        ssl_detected: false
```
Comment (asisayag2) [inline __TESTS__/unit/config/cloudinaryConfig.test.ts:41]: Didn't we say that's not relevant?

File: __TESTS__/unit/config/cloudinaryConfig.test.ts [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+    expect(conf.cloud.api_key).toBe('api_key');
+    expect(conf.cloud.api_secret).toBe('api_secret');
+    expect(conf.cloud.secure_cdn_subdomain).toBe(true);
+    expect(conf.cloud.secure).toBe(true);
+    expect(conf.cloud.secure_distribution).toBe(true);
+  });
+
+  it('Will populate URL configuration', () => {
+    const conf = new CloudinaryConfig({
+      url: {
```
Comment (asisayag2) [inline __TESTS__/unit/config/cloudinaryConfig.test.ts:57]: Missing 'analytics'

File: src/config/BaseConfig.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+class Config {
+  /**
+   * Type guard for user config input
+   * @param {any} object User provided input of unknown shape
+   * @param {string[]} validKeys Array of valid configuration keys
+   * @return {boolean} Whether an object is a valid configuration object
+   */
+  isValidConfig(object:unknown, validKeys:string[]): object is ICloudConfig {
+    if (typeof object !== 'object' || object instanceof Array) {
+      return false;
```
Comment (asisayag2) [inline src/config/BaseConfig.ts:13]: Sometimes you return 'false' sometimes you log an error ,why?

File: src/config/AccountConfig.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    super();
+    if (this.isAccountConfigTypeGuard(accountConfig)) {
+      Object.assign(this, accountConfig);
+    } else {
+      console.error(INVALID_TYPE_MESSAGE);
+    }
+  }
+
+  isAccountConfigTypeGuard(accountConfig:unknown): accountConfig is IAccountConfig {
+    return this.isValidConfig(accountConfig, ALLOWED_ACCOUNT_CONFIG);
```
Comment (asisayag2) [inline src/config/AccountConfig.ts:24]: Why not setting a member of the baseConfig with the ALLOWED values and let the baseConfig handle generically? 
Something like:
`this.validKeys = ALLOWED_ACCOUNT_CONFIG`

and in the `baseConfig` just use the `this.validKeys` in the `isValid` method

File: src/config/BaseConfig.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+import {INVALID_TYPE_MESSAGE} from "../constants";
+
+class Config {
+  /**
+   * Type guard for user config input
+   * @param {any} object User provided input of unknown shape
+   * @param {string[]} validKeys Array of valid configuration keys
+   * @return {boolean} Whether an object is a valid configuration object
+   */
+  isValidConfig(object:unknown, validKeys:string[]): object is ICloudConfig {
```
Comment (asisayag2) [inline src/config/BaseConfig.ts:11]: object?

File: src/config/URLConfig.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+class URLConfig extends Config implements IURLConfig {
+  force_version?: boolean;
+  shorten?: boolean;
+  sign_url?: boolean;
+  long_url_signature?: boolean;
+  use_root_path?: boolean;
+  auth_token?: IAuthTokenConfig;
+  [other:string]: unknown;
+
```
Comment (asisayag2) [inline src/config/URLConfig.ts:14]:
```suggestion
analytics?: boolean;
```

File: src/constants.ts [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+  'private_cdn',
+  'cdn_subdomain',
+  'secure_cdn_subdomain',
+  'cname',
+  'ssl_detected',
+  'secure'
+];
+
+
+export const INVALID_TYPE_MESSAGE = 'Invalid Type or key received';
```
Comment (asisayag2) [inline src/constants.ts:24]: Don't you want to say something about config?

File: src/config/BaseConfig.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+    if (typeof object !== 'object' || object instanceof Array) {
+      return false;
+    }
+
+    Object.keys(object).forEach((key) => {
```
Comment (asisayag2) [inline src/config/BaseConfig.ts:16]: Don't understand how this validation goes together with the `[other:string]: unknown;` in all interfaces

**Comment by patrick-tolosa** [comment]
@asisayag2  this was before the demo and updates. 

Please see this PR with all the new updates, I'm closing this PR for now.
https://github.com/cloudinary/cloudinary-js-base/pull/20

**Pull request closed without merging**